rotkehlchen/db/eth2.py

```diff
@@ -259,22 +259,45 @@ def add_or_update_validators_except_ownership(
             validators: list[ValidatorDetails],
     ) -> None:
         """Adds or update validator data but keeps the ownership already in the DB"""
-        pubkey_to_ownership = self.get_pubkey_to_ownership(write_cursor)
-        for entry in validators:
-            entry.ownership_proportion = pubkey_to_ownership.get(entry.public_key, ONE)
-        self.add_or_update_validators(write_cursor, validators)
+        self.add_or_update_validators(
+            write_cursor=write_cursor,
+            validators=validators,
+            updatable_attributes=('validator_index', 'withdrawal_address', 'activation_timestamp', 'withdrawable_timestamp'),  # noqa: E501
+        )
 
     def add_or_update_validators(
             self,
             write_cursor: 'DBCursor',
             validators: list[ValidatorDetails],
+            updatable_attributes: tuple[str, ...] = ('validator_index', 'ownership_proportion', 'withdrawal_address', 'activation_timestamp', 'withdrawable_timestamp'),  # noqa: E501
     ) -> None:
-        """Adds or updates validator data"""
-        write_cursor.executemany(
-            'INSERT OR REPLACE INTO '
-            'eth2_validators(validator_index, public_key, ownership_proportion, withdrawal_address, activation_timestamp, withdrawable_timestamp) VALUES(?, ?, ?, ?, ?, ?)',  # noqa: E501
-            [x.serialize_for_db() for x in validators],
-        )
+        """Adds or updates validator data
+
+        This used to do an INSERT OR REPLACE with an executemany. While that approach
+        was simpler and "worked" it also killed all foreign key relations and since
+        at the time of writing the daily stats have a foreign key relation to this table
+        all the daily stats were deleted.
+        """
+        for validator in validators:
+            result = write_cursor.execute(
+                'SELECT validator_index, public_key, ownership_proportion, withdrawal_address, '
+                'activation_timestamp, withdrawable_timestamp FROM eth2_validators '
+                'WHERE public_key=?', (validator.public_key,),
+            ).fetchone()
+            if result is not None:  # update case
+                db_validator = ValidatorDetails.deserialize_from_db(result)
+                for attr in updatable_attributes:
+                    if getattr(db_validator, attr) != (new_value := getattr(validator, attr)):
+                        write_cursor.execute(
+                            f'UPDATE eth2_validators SET {attr}=? WHERE public_key=?',
+                            (new_value, validator.public_key),
+                        )
+            else:  # insertion case
+                write_cursor.execute(
+                    'INSERT INTO '
+                    'eth2_validators(validator_index, public_key, ownership_proportion, withdrawal_address, activation_timestamp, withdrawable_timestamp) VALUES(?, ?, ?, ?, ?, ?)',  # noqa: E501
+                    validator.serialize_for_db(),
```

Comment:
sqlite has upsert that I believe is what you need here? https://www.sqlite.org/lang_upsert.html Mentioning it here just in case it could be useful

Reply:
upsert has some edge cases if you read and it's fairly new. After I started working on rotki ;).

But may work. Can try it later.